src/main/telemetry/crsf.c

```diff
@@ -337,20 +337,82 @@ void crsfFrameDeviceInfo(sbuf_t *dst) {
 }
 
 #if defined(USE_CRSF_CMS_TELEMETRY)
+#define CRSF_DISPLAYPORT_MAX_CHUNK_LENGTH   50
+#define CRSF_DISPLAYPORT_BATCH_MAX          0x3F
+#define CRSF_DISPLAYPORT_FIRST_CHUNK_MASK   0x80
+#define CRSF_DISPLAYPORT_LAST_CHUNK_MASK    0x40
+#define CRSF_DISPLAYPORT_SANITIZE_MASK      0x60
+#define CRSF_RLE_CHAR_REPEATED_MASK         0x80
+#define CRSF_RLE_MAX_RUN_LENGTH             256
+#define CRSF_RLE_BATCH_SIZE                 2
+
+static size_t getRunLength(const void *start, const void *end)
+{
+    uint8_t *cursor = (uint8_t*)start;
+    uint8_t *endPtr = (uint8_t*)end;
+    uint8_t c = *cursor;
+    size_t runLength = 0;
+    for (; cursor != endPtr; cursor++) {
+        if (*cursor == c) {
+            runLength++;
+        } else {
+            break;
+        }
+    }
+    return runLength;
+}
+
+static sbuf_t *cRleEncodeStream(sbuf_t *source, sbuf_t *dest, size_t maxDestLen)
+{   
+    const uint8_t *destEnd = sbufPtr(dest) + maxDestLen;
+    const uint8_t *srcEnd = source->end;
+    while (sbufBytesRemaining(source) && (sbufPtr(dest) < destEnd)) {
+        const uint8_t destRemaining = destEnd - sbufPtr(dest);
+        const uint8_t *srcPtr = sbufPtr(source);
+        const size_t runLength = getRunLength(srcPtr, srcEnd);
+        uint8_t c = *srcPtr;
+        c = (c & CRSF_DISPLAYPORT_SANITIZE_MASK) ? c : ' ';  //sanitize input
```

Comment:
This seems to be somewhat problematic, as it makes half of the character set unavailable for CMS - have you investigated how much extra overhead we'd be getting by using an extra character (like 0xff) to indicate a RLE encoded sequence?

Reply:
This was decided after deeply inspecting the compatibility between the character sets.  The font in CMS is highly customized, so there is only a partial region of characters that will align on the client/lua side.  Everything below 32 is non-printable, and everything above 127 is high ascii and does not match the BF font set.  It wouldn't add a ton of length to the data by adding extra characters ahead of run lengths, but this was kind of a good combination for this particular use case.  An alternative might be to flag the runlength with 0x80 and create a max run length of 127.   This would open up the entire byte for encoding, but could create additional run length batches when they exceed 127 chars.

Reply:
to me, this felt like a good compromise with little waste

Reply:
Also -- defaulting incompatible chars to 32 save cpu cycles during rendering on the client side.  Without skipping characters, the remote hard fails with CPU limit exceeded.

Reply:
Good point - we need the characters used in CMS to be restricted to characters for which the Betaflight MAX7456 font matches ASCII.
We might even want to push this down to a deeper layer of CMS, as non-matching characters will break CMS for other places that use it, like Spektrum or the dashboard.
But if your intention is to substitute characters > 127 as well then the current filtering will not work. Probably easier to do as `if (c < 0x20 || c >= 0x60) { c = ' '; }` to capture the range of (mostly) matching characters.

Reply:
![image](https://user-images.githubusercontent.com/8164265/80853724-fbc37200-8c00-11ea-9d97-f0741288a29b.png)

Reply:
![image](https://user-images.githubusercontent.com/8164265/80853938-675a0f00-8c02-11ea-82e4-8d2d7a96990a.png)
Green = Can be safely interpreted on the client side as seen in the table in the previous comment, but can't be shown properly on the OSD.

Reply:
@codecae: None of the ones above 0x5f matches:
![image](https://user-images.githubusercontent.com/4742747/80853949-8f8c5100-8c88-11ea-9ef2-5637cddc3f0a.png)

Also, 0x24 does not match either. :confounded:

Reply:
> Green = Can be safely interpreted on the client side as seen in the table in the previous

Not sure what you mean by 'can be interpreted' - these do not match the respective ASCII characters, and most of them also don't have good matches with any ASCII characters - using them in CMS will result in garbled display output.

Reply:
opentx font
![image](https://user-images.githubusercontent.com/8164265/80854384-bfdedb80-8c05-11ea-8ee8-83fa836449d9.png)
![image](https://user-images.githubusercontent.com/8164265/80854393-d8e78c80-8c05-11ea-85e0-9be2b7c94ce9.png)

Reply:
As I said, anything above 0x59 does not match the characters of the Betaflight MAX7456 font that CMS uses.

Reply:
Re: "can be interpreted"  -- there are cases in CMS where lowercase values are used (ie: 'noRevis' on uncommitted builds) ... The lua script has no problem displaying these lowercase values but the CMS screen does not.

I'll plant a better filter into CMS and remove it from here.  Any opposition to replacing these characters with spaces?

Reply:
Thanks. Replacing non-printables with spaces sounds like the best option - even if we get a character that is a lowercase letter in ASCII we won't know if it was intended to be that, or intended to be the MAX7456 character.

Reply:
See also #9751 - lowercase characters in CMS simply don't work.